tools/src/ops_tools/_update_charmcraft_yaml.py

```diff
@@ -0,0 +1,191 @@
+
+from __future__ import annotations
+
+import argparse
+import difflib
+import importlib
+import re
+import sys
+from typing import Any, Generator
+
+import yaml
+
+from . import ActionDict, OptionDict, action_to_juju_schema, config_to_juju_schema
+
+
+def get_class_from_module(class_specifier: str) -> Generator[type]:
+    """Import the specified module and get the class from the top-level namespace."""
+    if ':' in class_specifier:
+        module_name, class_name = class_specifier.rsplit(':', 1)
+    else:
+        module_name = 'src.charm'
+        class_name = class_specifier
+    module = importlib.import_module(module_name)
+    for attr in dir(module):
+        if not isinstance(getattr(module, attr), type):
+            continue
+        if re.fullmatch(class_name, attr):
+            yield getattr(module, attr)
+
+
+def _insert_into_charmcraft_yaml(
+    raw_yaml: str, section_name: str, replacement: dict[str, Any]
+) -> str:
+    """Surgically insert a section into charmcraft.yaml.
+
+    The specified section of charmcraft.yaml is replaced with the provided data.
+    However, the rest of the file is left unchanged; in particular, comments and
+    ordering is preserved.
+    """
+    # To simplify the regular expressions, look for four variants. Firstly,
+    # there is a section with YAML both before and after it.
+    mo = re.match(
+        rf'(?P<before>.*)^{section_name}:(?P<tab_size>\s+).+?^(?P<after>\w.*)',
+        raw_yaml,
+        re.DOTALL | re.MULTILINE,
+    )
+    if mo:
+        replacement_section = yaml.safe_dump(replacement, indent=len(mo['tab_size']))
+        return f'{mo["before"]}{replacement_section}{mo["after"]}'
+    # Secondly, there is a section with YAML before it, but no section after it.
+    mo = re.match(
+        rf'(?P<before>.*)^{section_name}:(?P<tab_size>\s+).+', raw_yaml, re.DOTALL | re.MULTILINE
+    )
+    if mo:
+        replacement_section = yaml.safe_dump(replacement, indent=len(mo['tab_size']))
+        return f'{mo["before"]}{replacement_section}'
+    # Next, there is a section with no YAML before it, but YAML after it.
+    mo = re.match(
+        rf'^{section_name}:(?P<tab_size>\s+).+?^(?P<after>\w.*)',
+        raw_yaml,
+        re.DOTALL | re.MULTILINE,
+    )
+    if mo:
+        replacement_section = yaml.safe_dump(replacement, indent=len(mo['tab_size']))
+        return f'{replacement_section}{mo["after"]}'
+    # Finally, there is no existing config section.
+    return f'{raw_yaml}\n{yaml.safe_dump(replacement, indent=2)}'
+
+
+def main():
+    """Merge generated config and action sections into charmcraft.yaml."""
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument(
+        '--charmcraft-yaml',
+        help='Path to the charmcraft.yaml file to update.',
+        default='charmcraft.yaml',
+    )
+    parser.add_argument(
+        '--config-class',
```

Comment:
I realise this is more accurate, but it's a bit long. Would `--config` and `--action` be enough?